src/build.cc

```diff
@@ -628,45 +684,54 @@ bool Builder::Build(string* err) {
   // command runner.
   // Second, we attempt to wait for / reap the next finished command.
   while (plan_.more_to_do()) {
-    // See if we can start any more commands.
-    if (failures_allowed && command_runner_->CanRunMore()) {
-      if (Edge* edge = plan_.FindWork()) {
-        if (edge->GetBindingBool("generator")) {
+    // See if we can start any more commands...
+    bool can_run_more =
+        failures_allowed   &&
+        plan_.more_ready() &&
+        command_runner_->CanRunMore();
```

Comment:
This `CanRunMore()` is kind of wrong.  If there is a token pool, it's the *sole* authority on parallelism.

If you have an `AcquireToken()` now, then you should only be calling `CanRunMore()` as an implementation detail within `AcquireToken`  in the fallback case when a token pool is absent.